### DISCUSSION TITLE
What is the correct way to extend a leaflet class, such as Marker, while using TypeScript.

### DISCUSSION THREAD

**Discussion by MLaidlawScott** [Leaflet/Leaflet#9367] [Q&A] [answered]
Hello

I have read the tutorial on [extending the class system](https://leafletjs.com/examples/extending/extending-1-classes.html) and have managed to extend Marker with my desired functionality. However, I am unable to get everything working with no TypeScript errors. 

I have my marker
`const PhysicalSizeMarker = L.Marker.extend({
  options: {
    width: 0,
    height: 0,
  }, ... })
`

and I have my factory

```
export function physicalSizeMarker(
  latlng: L.LatLngExpression,
  options: PhysicalSizeMarkerOptions,
) {
  return new PhysicalSizeMarker(latlng, options);
}
```

This gives me, for the constructor:
`Expected 0 arguments, but got 2`.

I can make these errors go away by extending the Marker class in the usual JavaScript way:

```
class PhysicalSizeMarker extends L.Marker {
  constructor(latlng: L.LatLngExpression, options: PhysicalSizeMarkerOptions) {
    super(latlng, options);
  }
```

But your tutorial says that .extend should be used to extend classes.

How can I type my constructor while extending a marker?

Thanks

**Comment by chrisnruud** [comment]
I am also interested in this, because it seems like the .extend and factory methods are using a legacy approach that served well when JS was at the stage it was when the system was first introduced, but TS is way more Object Oriented and it feels dirty and wrong to bypass implicit instructions from maintainer , just to make it compatible with the typing system.

**Comment by cyclingbyte** [comment]
To do this, you need to write object-oriented js. Furthermore, you have to declare the leaflet module with `declare module 'leaflet' {}`
I just tried this with an exiting plugin. Compare https://github.com/cyclingbyte/Leaflet.SidePanel with https://github.com/cyclingbyte/Leaflet.SidePanel/tree/6-add-ts
I'm not 100% finished, but everything looks working at the moment.
(Unfortunately I change the prettier, so there are some unnecessary changes in the PR https://github.com/cyclingbyte/Leaflet.SidePanel/pull/16)

Edit: Maybe the object-oriented js isn't necessary, and the declaration are enough...?

**Comment by mrrizwan304** [comment]
To extend a Leaflet class like Marker using TypeScript, follow these steps:

Create a New Class: Extend the Marker class.
Use TypeScript Annotations: Ensure proper typings by importing Marker from Leaflet.
typescript
Copy code
```
import { Marker, LatLngExpression } from 'leaflet';

class CustomMarker extends Marker {
    constructor(latlng: LatLngExpression, options?: L.MarkerOptions) {
        super(latlng, options);
    }

    // Add custom methods or override existing ones
}
```
This setup lets you create a new class extending Marker with TypeScript's type safety.

**Reply by mrrizwan304** [reply]
To extend a Leaflet class like Marker using TypeScript, create a new class that extends Marker, ensuring proper typings with TypeScript annotations. This approach, similar to the efficient methods we use at[ sell my house fast carrollton georgia](https://www.urgenthomeselling.com/we-buy-houses-carrollton-georgia/), provides type safety while allowing custom methods or overrides.

**Reply by MLaidlawScott** [reply]
I was able to get this to work too but the leaflet documentation explicitly says to use L.marker.extend. If it works though I don't see why not to use this approach.

**Reply by Falke-Design** [reply, marked as answer]
With Leaflet V2 `class CustomMarker extends Marker` will be the prefered way
